src/packer.ts

```diff
@@ -0,0 +1,195 @@
+    private options: Partial<PackerOptions>;
+    private len: number; // Total number of values
+    private idx: Record<string, IndexValues>;
+    private shape: Record<string, number[]>; // shape of each named value group i.e. size of each dimension
+
+    constructor(options: Partial<PackerOptions>) {
+        this.options = options;
+
+        this.idx = {};
+        this.shape = {};
+
+        const { scalar, fixed } = options;
+        const arrayOptions = options.array;
+
+        const allNames = [...(scalar || []), ...(arrayOptions?.keys() || []), ...(fixed?.keys() || [])];
+        // Check for duplicate names
+        const dup = allNames.find((name, i) => allNames.lastIndexOf(name) !== i)
+        if (dup) {
+            // TODO: make a throw error util
+            throw Error(`Names must be distinct between 'scalar', 'array' and 'fixed'. ${dup} appears in more ` +
+                        "than one place.");
+        }
+
+        if (scalar) {
+            Array.from(scalar).forEach((name, i) => {
+                // for each scalar, shape gets an array with a single integer in it, set to 0 (indicating no dimensions)
+                this.shape[name] = [0];
+                // .. and index gets the index where they can be found
+                this.idx[name] = { start: i, length: 1};
+            });
+        }
+
+        this.len = scalar?.size || 0;
+        if (arrayOptions) {
+            for (const [name, value] of arrayOptions) {
+                const tmp = preparePackArrayForShape(name, value);
+                this.shape[name] = tmp.shape;
+                // array of 0-based indexes where each of the named values can be found
+                this.idx[name] = { start: this.len, length: tmp.n }
+                this.len = this.len + tmp.n;
+            }
+        }
+
+        if (!this.len) {
+            throw Error("Trying to generate an empty packet. You have not provided any entries in 'scalar' or" +
+                        " 'array', which implies generating from a zero-length parameter vector.");
+        }
+    }
+
+    // TODO: unpack_array and unpack_ndarray are currently separate methods, but could be combined as in the R
+    // implementation.
+    // TODO: The return types for these methods are a bit of a mess - handing back a combination of scalars, ordinary
+    // array and ndarrays for the unpack_array method, ndarrays only for unpack_ndarray. We should probably EITHER
+    // only return ndarrays for any array results OR always return oridinary arrays for 1D results.
+
+    // Unpack a one-dimensional array
+    public unpack_array(x: Array<number>) {
+        if (x.length !== this.len) {
+            throw Error(`Incorrect length input; expected ${this.len} but given ${x.length}.`);
+        }
+
+        // Return a map of names to values in the format described by shape
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        let result = new Map<string, any>();
```

Comment:
`Map<string, number | number[] | ndarray>`  rather than any?

Returning `number | ndarray` would probably be sensible

Reply:
Yep, let's give back numbers for scalars and ndarray for everything else.